plan/rule_partition_prunner.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+// // Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/table/tables"
+	"github.com/pingcap/tidb/util/ranger"
+)
+
+// partitionPrunner rewrites the ast for table partition.
+//
+// create table t (id int) partition by range (id)
+//   (partition p1 values less than (10),
+//    partition p2 values less than (20),
+//    partition p3 values less than (30))
+//
+// select * from t is equal to
+// select * from (union all
+//      select * from p1 where id < 10
+//      select * from p2 where id < 20
+//      select * from p3 where id < 30)
+//
+// partitionPrunner is here because it's easier to prune partition after predicate push down.
+type partitionPrunner struct{}
```

Comment:
It does more than partition pruning, including rewriting data source.
`partitionProcessor` is more appropriate.

Reply:
So, what do you think? @shenli @zz-jason